8519: universal search design proposals (#9324)
CDCgov/prime-reportstream: prime-router/docs/proposals/0020-universal-metadata-search/0020-universal-metadata-search.md

            }
          ],
          "tag": [
            {
              "system": "http://terminology.hl7.org/CodeSystem/v2-0103",
              "code": "P"
            }
          ]
        },
        "identifier": [
          { "value": "DCEED.ELR" },
          {
            "type": {
              "coding": [
                {
                  "system": "http://terminology.hl7.org/CodeSystem/v2-0301",
                  "code": "ISO"
                }
              ]
            },
            "value": "urn:oid:2.16.840.1.114222.4.1.144.2.5"
          }
        ]
      }
    }
```

Pros:

- Exposes a REST API so developers don't need to learn a new syntax
- Has excellent support for fuzzy searches
- Automatically indexes the data
- Azure has a fully hosted version

Cons:

- Does not work with FHIR out of the box
- Linking back to the lineage data in postgres will be complicated, expensive and would not potentially scale to all use cases
- Does not fully support inserting nested objects
- Can be expensive

#### Store the data in a normalized form

While we could roll our own version of this (see [this](https://github.com/CDCgov/prime-reportstream/pull/7228) proposal),

> there is an existing solution in the open source [HAPI FHIR server](https://hapifhir.io/hapi-fhir/docs/) that could be

Author: ```suggestion
there is an existing solution in the open source [HAPI FHIR server](https://hapifhir.io/hapi-fhir/docs/server_jpa/introduction.html) (I tested the JPA server, but the plain one could be an option as well depending on the requirements) that could be
```

# Proposal for Universal Metadata Search

## Goals

- Enable the existing search needs for a variety of users
  - Receivers
  - Admins
  - Senders

> - These users need to perform queries against report and metadata across different dimensions including

Author: I was trying to refer to the actual reports that get sent through the system, but I think that might be confusing alongside the metadata reference.

> - These users need to perform queries against report and metadata across different dimensions including

Reviewer: Ohhhh you meant report like the data type not the reports that are wanted from these queries, got it. I was confused.

  - date ranges
  - aggregates

See the initial [requirements](../0018-universal-search.md).

The search question in that document breakdown into a few broad categories:

> - information about one or more report(s) that is queryable by dates, sender and receiver

Author: ```suggestion
- information about one or more report(s) that is queryable by dates, sender, receiver and status
```

- information about one or more item(s) that is queryable by metadata fields
- aggregates around one or more item(s) 

## Problems to solve for

- Report and item lineages are tough to query
- Querying and aggregating metadata from different result types as the system is enhanced

## Potential Solutions

### Enhancing the ability to query reports and items

The first problem to solve for is how to efficiently query how items flow from a sender to a receiver.
This is currently captured via the report and item lineages; this is a graph structure where each item and report 
has a link to its parent.  Some example questions that a solution would need to solve for is:
 
> As a sender, what are the list of reports I sent out grouped by receiver?

> > As a receiver, what are the reports were sent to me in a date range? 

Author: ```suggestion
> As a receiver, what are the reports that were sent to me in a date range? 
```

#### Store the origin details on each child item and create a new table that links an origin report to all delivered items

There is a current solution where each child item contains a link to the origin report 
(`origin_report_id`/`origin_report_index`) that enables quickly starting at the bottom of a lineage graph and jumping 

> back up to the original source.  A similar method could be done for going from an original report to it's terminal items

Author: ```suggestion
back up to the original source.  A similar method could be done for going from an original report to its terminal items
```

        ItemLineage.ITEM_LINEAGE.PARENT_INDEX,
    )
        .from(ItemLineage.ITEM_LINEAGE)
        .where(ItemLineage.ITEM_LINEAGE.CHILD_REPORT_ID.`in`(reportIds))
        .unionAll(
            select(
                ItemLineage.ITEM_LINEAGE.PARENT_REPORT_ID,
                ItemLineage.ITEM_LINEAGE.PARENT_INDEX,
            )
                .from(table(name("t")))
                .join(ItemLineage.ITEM_LINEAGE)
                .on(
                    field(name("t", "parent_report_id"), UUID)
                        .eq(ItemLineage.ITEM_LINEAGE.PARENT_REPORT_ID),
                    field(name("t", "parent_index"), INTEGER)
                        .eq(ItemLineage.ITEM_LINEAGE.PARENT_INDEX)
                )
        )
)

val lineageQuery = create.withRecursive(cte).selectFrom(cte)
val lineageWithinFiveDays = lineageQuery.where(
    ItemLineage.ITEM_LINEAGE.CREATED_AT.between(
        OffsetDateTime.now().minusDays(5),
        OffsetDateTime.now()
    )
)
```

Pros:

- Allows writing queries against any part of the lineage hierarchy
- Doesn't require capturing any new data or adjusting how report stream works
- Is a very flexible solution for future needs

Cons:

> - Developers need to understand how recursive queries

Author: ```suggestion
- Developers need to understand how recursive queries work
```

- Indices can be written for the JSONB data to support fast queries
- JOOQ supports JSONB operators out of the box
- Joining back to the report, item and lineage data is easy

Cons:

- Engineers will need to have a good understanding of the postgres JSONB type
- Does not scale to arbitrary searching of the metadata
- Individual searches will need to be tuned to make sure an effective index is chosen
- Inserts will get progressively more expensive as more and more indices are added 

#### Mongodb

The FHIR bundle could be inserted directly into a mongo collection and then indices would be added to support the various
use cases.

Pros:

- Easily supports new FHIR resource types without any changes
- Allows splitting the data across multiple nodes easily
- Has good support for writing aggregate queries

Cons:

- Requires an infrastructure change
- If the lineage data is still stored in postgres, requires selecting IDs and then doing a large `in` query
- Developers might not necessarily be familiar with MongoDB
- Concurrent updates will be complicated; i.e. the send function fanning out and making multiple updates to the same
original report

#### Elasticsearch

> An excellent option from a search perspective as it handles indexing unstructured date extremely effificently, but has 

Author: ```suggestion
An excellent option from a search perspective as it handles indexing unstructured data extremely efficiently, but has 
```

          "extension": [
            {
              "url": "http://ibm.com/fhir/cdm/StructureDefinition/source-data-model-version",
              "valueString": "2.5.1"
            }
          ],
          "tag": [
            {
              "system": "http://terminology.hl7.org/CodeSystem/v2-0103",
              "code": "P"
            }
          ]
        },
        "identifier": [
          { "value": "DCEED.ELR" },
          {
            "type": {
              "coding": [
                {
                  "system": "http://terminology.hl7.org/CodeSystem/v2-0301",
                  "code": "ISO"
                }
              ]
            },
            "value": "urn:oid:2.16.840.1.114222.4.1.144.2.5"
          }
        ]
      }
    }
```

Pros:

- Exposes a REST API so developers don't need to learn a new syntax
- Has excellent support for fuzzy searches
- Automatically indexes the data
- Azure has a fully hosted version

Cons:

- Does not work with FHIR out of the box
- Linking back to the lineage data in postgres will be complicated, expensive and would not potentially scale to all use cases
- Does not fully support inserting nested objects
- Can be expensive

> - 

Author: ```suggestion
```

#### Store the data in a normalized form

While we could roll our own version of this (see [this](https://github.com/CDCgov/prime-reportstream/pull/7228) proposal),

> there is an existing solution in the open source [HAPI FHIR server](https://hapifhir.io/hapi-fhir/docs/) that could be

Author: I'm not necessarily following which link you're looking for?

> there is an existing solution in the open source [HAPI FHIR server](https://hapifhir.io/hapi-fhir/docs/) that could be

Reviewer: Documentation more specific to the open source solution you are talking about here. Those just link to the braod HAPI library docs, but not to s specific part of them that talks about the solution, if that makes sense.

> there is an existing solution in the open source [HAPI FHIR server](https://hapifhir.io/hapi-fhir/docs/) that could be

Author: Got it!

# Proposal for Universal Metadata Search

## Goals

- Enable the existing search needs for a variety of users
  - Receivers
  - Admins
  - Senders
- These users need to perform queries against report and metadata across different dimensions including
  - date ranges
  - aggregates

See the initial [requirements](../0018-universal-search.md).

> The search question in that document breakdown into a few broad categories:

Author: ```suggestion
The search question in that document break down into a few broad categories:
```

- information about one or more report(s) that is queryable by dates, sender, receiver and status
- information about one or more item(s) that is queryable by metadata fields
- aggregates around one or more item(s) 

## Problems to solve for

- Report and item lineages are tough to query
- Querying and aggregating metadata from different result types as the system is enhanced

## Potential Solutions

### Enhancing the ability to query reports and items

The first problem to solve for is how to efficiently query how items flow from a sender to a receiver.
This is currently captured via the report and item lineages; this is a graph structure where each item and report 

> has a link to its parent.  Some example questions that a solution would need to solve for is:

Author: ```suggestion
has a link to its parent.  Some example questions that a solution would need to solve for:
```

- Can be expensive

#### Store the data in a normalized form

While we could roll our own version of this (see [this](https://github.com/CDCgov/prime-reportstream/pull/7228) proposal),
there is an existing solution in the open source [HAPI FHIR server](https://hapifhir.io/hapi-fhir/docs/server_jpa/introduction.html) (I tested the JPA server, but the plain one could be an option as well depending on the requirements) that could be
leveraged to accomplish the same goal.  This solution supports normalizing any FHIR resource type and is a more extensible
solution than something that is custom.

This is an open source server that accepts FHIR data and then provides various mechanisms for searching that data and
could serve as the tool used for querying against the items that flow through ReportStream.

![search](./search-patient.png)
![results](./search-results.png)
![db](./resource-in-db.png)

This approach would entail spinning up a HAPI FHIR server and then accessing the data via the defined API endpoints

Pros:

- Is relatively easy to get started
- Out of the box search functionality is robust
- Provides fantastic support for arbitrary searches

Cons:

> - Out of the box, `type: "message"` was not accepted by the server; we would likely need to fork or at least figure out to

Reviewer: [how] to get it accepted

> - Out of the box, `type: "message"` was not accepted by the server; we would likely need to fork or at least figure out to

Author: ```suggestion
- Out of the box, `type: "message"` was not accepted by the server; we would likely need to fork or at least figure out how to
```

definitely interested in knowing around how data is getting routed (this is also what someone debugging the system is
interested in), but items such as what is valuable to search in the metadata is still an unknown.

Due to the ambiguity, the recommended approach is:

- track the origin report/items to the terminal report/items in a new column and table
  - Specifically, we would insert a new row whenever there is no next action to be performed
- Create a new metadata table with a JSONB column for storing the metadata
  - After creating the internal FHIR bundle that the UP will operate on, insert that into the metadata table with 
  origin report and index

This is the simplest approach and requires no new infrastructure or concepts while still maintaining flexibility moving
forward as requirements evolve.  The largest cons of this solution are mostly ones of scale and monitoring could be
configured such that the team would have an early warning that one of the other more involved solutions spelled out 
here need to be adopted

## Searches to be split into other tickets

> - Returning a list of "undelivered" reports

Reviewer: I worry about us not tackling the delivered/undelivered aspect in this ticket. It seems like we could put a lot of work going down one route just to do this research and decide another makes more sense.

Author: I'll update this in the doc too, but delivered reports is definitely in scope and covered by this proposal.  I think agree and will include capturing "undelivered" reports too.  The original reason to separate that out was that this data does get captured in the `action` table (action is `send_warning`), but the report id is in a raw string in the action param and results and not easily searchable.  We would need to figure out how to adjust the schema to surface that info in a more queryable fashion
